### PR TITLE
feat(cloudformation): add an option to override parameters

### PR DESCRIPTION
**Pull request by nikpivkin**
See https://github.com/aquasecurity/trivy/issues/5167

It is currently possible to pass variables for terraform and helm, but not for CloudFormation. This PR adds the ability to override parameters in CloudFormation templates. Then in Trivy we can add flags to pass parameters as a key-value list or path to the file that contains the parameters (like the `--parameter-overrides` flag in `aws-cli`).

Ref:
- https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudformation/deploy.html#deploy
- https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudformation/deploy.html#supported-json-syntax

### CODE REVIEW

File: pkg/scanners/cloudformation/parser/parameter.go [hunk 1]
```diff
@@ -57,3 +60,67 @@ func (p *Parameter) UpdateDefault(inVal interface{}) {
...
+
+type Parameters map[string]any
+
+func (p *Parameters) Merge(other Parameters) {
+	for k, v := range other {
+		(*p)[k] = v
+	}
+}
+
+func (p *Parameters) UnmarshalJSON(data []byte) error {
```
Comment (simar7) [inline pkg/scanners/cloudformation/parser/parameter.go:72]: What about YAML? Should we add support for that too?

File: pkg/scanners/cloudformation/parser/parameter.go [hunk 2]
```diff
@@ -57,3 +60,67 @@ func (p *Parameter) UpdateDefault(inVal interface{}) {
...
+
+			// CloudFormation like format
+			var cfparams []struct {
+				ParameterKey   string `json:"ParameterKey"`
+				ParameterValue string `json:"ParameterValue"`
+			}
+
+			if err := json.Unmarshal(data, &cfparams); err != nil {
+				return err
+			}
```
Comment (simar7) [inline pkg/scanners/cloudformation/parser/parameter.go:115]: There's a note in the docs that says:

> Only ParameterKey and ParameterValue are expected keys, command will throw an exception if receives unexpected keys (e.g. UsePreviousValue or ResolvedValue).

I tried with an input such as:
```json
[
    {
        "UsePreviousValue": "Key1",
        "ResolvedValue": "Value1"
    }
]
```

And expectedly I got nothing unmarshaled back, which is what I expected. 

But I'm curious silently missing parsing will end up ignoring some errors down the codepath as the AWS SDK says that it currently throws an exception on such a case. What do you think, should we improve the handling or its fine?

File: pkg/scanners/cloudformation/parser/parser.go [hunk 3]
```diff
@@ -151,8 +189,48 @@ func (p *Parser) ParseFile(ctx context.Context, fs fs.FS, path string) (context
...
+		r.ConfigureResource(name, fsys, path, context)
 	}
 
 	return context, nil
 }
+
+func (p *Parser) parseParams() error {
+	if p.overridedParameters != nil { // parameters have already been parsed
+		return nil
+	}
```
Comment (simar7) [inline pkg/scanners/cloudformation/parser/parser.go:201]: when would this case happen? just curious.

File: pkg/scanners/cloudformation/parser/parameter.go (hunk 1 above)
Comment (nikpivkin) [reply pkg/scanners/cloudformation/parser/parameter.go:72]: That's right, AWS only supports json, so users will use json as well.

File: pkg/scanners/cloudformation/parser/parameter.go (hunk 2 above)
Comment (nikpivkin) [reply pkg/scanners/cloudformation/parser/parameter.go:115]: I'll add a check

File: pkg/scanners/cloudformation/parser/parser.go (hunk 3 above)
Comment (nikpivkin) [reply pkg/scanners/cloudformation/parser/parser.go:201]: When parsing the file system, the parameters will be parsed once, not before parsing each CF template.

File: pkg/scanners/cloudformation/parser/parameter.go [hunk 4]
```diff
@@ -110,7 +111,9 @@ func (p *Parameters) UnmarshalJSON(data []byte) error {
 				ParameterValue string `json:"ParameterValue"`
 			}
 
-			if err := json.Unmarshal(data, &cfparams); err != nil {
+			d := json.NewDecoder(bytes.NewReader(data))
+			d.DisallowUnknownFields()
```
Comment (simar7) [inline pkg/scanners/cloudformation/parser/parameter.go:115]: neat! I didn't know about this

**Review by simar7 — APPROVED** [review]

**Pull request merged**